### ISSUE TITLE
Have to run the pkill command twice to kill kytos foreground process

### ISSUE THREAD

**Issue by ajoaoff** [kytos-ng/kytos#113]
Original issue opened by @rmotitsuki at https://github.com/kytos/kytos/issues/1236.

When executing kytosd in foreground mode the user must run the command `pkill kytosd` twice.
It seems that the kytos terminal runs in a different thread and it is not cancelled.

**Comment by italovalcy** [comment]
This issue is related to Issue #129 and PR #130. Basically, after applying the PR 130, the issue of having to run pkill twice started happening again.

**Comment by italovalcy** [comment]
> This issue is related to Issue #129 and PR #130. Basically, after applying the PR 130, the issue of having to run pkill twice started happening again.

If we revert the PR 130, Kytosd in foreground we not provide the controller variable and also it will still not stop after the first Kill. Here is an execution example using the latest docker image (amlight/kytos:latest - 41434b0afd9b):

```
# Terminal 1
root@90aee3fe58e8:/# kytosd -f
Web update - Web UI was not updated

      _          _
     | |        | |
     | | ___   _| |_ ___  ___
     | |/ / | | | __/ _ \/ __|
     |   <| |_| | || (_) \__ \
     |_|\_\__,  |\__\___/|___/
            __/ |
           |___/

    Welcome to Kytos SDN Platform!

    We are making a huge effort to make sure that this console will work fine
    but for now it's still experimental.

    Kytos website.: https://kytos.io/
    Documentation.: https://docs.kytos.io/
    OF Address....: tcp://0.0.0.0:6653
    WEB UI........: http://0.0.0.0:8181/
    Kytos Version.: 2022.1rc1

 * Serving Flask app "kytos.core.controller" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
kytos $> controller
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
/usr/lib/python3.9/concurrent/futures/thread.py in <module>
----> 1 controller

NameError: name 'controller' is not defined

kytos $>


# Terminal 2:
root@90aee3fe58e8:/# pkill kytosd && sleep 5 && ps aux | grep kytos
root@90aee3fe58e8:/# ps aux | grep kytos
root       165  3.7  2.0 672960 81400 pts/0    Sl+  00:47   0:02 /usr/bin/python3 /usr/local/bin/kytosd -f
root       200  0.0  0.0   3176   660 pts/1    S+   00:48   0:00 grep kytos

# Terminal 1 (I've just pressed ENTER):
kytos $>
Stopping Kytos daemon... Bye, see you!
```

So, in the above example, Kytos' console keeps waiting on an additional ENTER to finishes.

**Comment by italovalcy** [comment]
The following patch seems to close the console when the first `pkill` is received. However, it leaves the terminal in an inconsistent state (you have to run `reset`)

```
--- /usr/local/lib/python3.9/dist-packages/kytos/core/kytosd.py.orig	2022-01-19 10:04:27.000000000 +0000
+++ /usr/local/lib/python3.9/dist-packages/kytos/core/kytosd.py	2022-01-20 18:16:11.898125016 +0000
@@ -35,7 +35,7 @@
         os.chmod(pid_dir, 0o1777)  # permissions like /tmp


-def start_shell(controller=None):
+def create_shell(controller=None):
     """Load Kytos interactive shell."""
     kytos_ascii = r"""
       _          _
@@ -82,9 +82,10 @@

     ipshell = InteractiveShellEmbed(config=cfg,
                                     banner1=banner1,
-                                    exit_msg=exit_msg)
+                                    exit_msg=exit_msg,
+                                    user_ns={'controller': controller})
     ipshell.prompts = KytosPrompt(ipshell)
-    ipshell()
+    return ipshell


 # def disable_threadpool_exit():
@@ -129,12 +130,15 @@

     async def start_shell_async():
         """Run the shell inside a thread and stop controller when done."""
-        _start_shell = functools.partial(start_shell, controller)
+        interactive_shell = create_shell(controller)

         try:
-            data = await loop.run_in_executor(executor, _start_shell)
+            data = await loop.run_in_executor(executor, interactive_shell)
         finally:
+            # Exit all embedded code in shell
             stop_controller(controller)
+            interactive_shell.magic("%kill_embedded -x -y")
+            interactive_shell.pt_loop.stop()
         return data

     loop = asyncio.get_event_loop()
```

Maybe we could send a signal to the interactive shell to represent Ctrl-D.

**Comment by ajoaoff** [comment]
> The following patch seems to close the console when the first `pkill` is received. However, it leaves the running shell in an inconsistent state (you have to run `reset`)
> 
> ```
> --- /usr/local/lib/python3.9/dist-packages/kytos/core/kytosd.py.orig	2022-01-19 10:04:27.000000000 +0000
> +++ /usr/local/lib/python3.9/dist-packages/kytos/core/kytosd.py	2022-01-20 18:16:11.898125016 +0000
> @@ -35,7 +35,7 @@
>          os.chmod(pid_dir, 0o1777)  # permissions like /tmp
> 
> 
> -def start_shell(controller=None):
> +def create_shell(controller=None):
>      """Load Kytos interactive shell."""
>      kytos_ascii = r"""
>        _          _
> @@ -82,9 +82,10 @@
> 
>      ipshell = InteractiveShellEmbed(config=cfg,
>                                      banner1=banner1,
> -                                    exit_msg=exit_msg)
> +                                    exit_msg=exit_msg,
> +                                    user_ns={'controller': controller})
>      ipshell.prompts = KytosPrompt(ipshell)
> -    ipshell()
> +    return ipshell
> 
> 
>  # def disable_threadpool_exit():
> @@ -129,12 +130,15 @@
> 
>      async def start_shell_async():
>          """Run the shell inside a thread and stop controller when done."""
> -        _start_shell = functools.partial(start_shell, controller)
> +        interactive_shell = create_shell(controller)
> 
>          try:
> -            data = await loop.run_in_executor(executor, _start_shell)
> +            data = await loop.run_in_executor(executor, interactive_shell)
>          finally:
> +            # Exit all embedded code in shell
> +            interactive_shell.magic("%kill_embedded -x -y")
> +            interactive_shell.pt_loop.stop()
>              stop_controller(controller)
>          return data
> 
>      loop = asyncio.get_event_loop()
> ```
> 
> Maybe we could send a signal to the interactive shell to represent Ctrl-D.

Does this patch fixes the issue with the `controller` variable being unavailable? Without it the console is almost useless.

**Comment by italovalcy** [comment]
> Does this patch fixes the issue with the `controller` variable being unavailable? Without it the console is almost useless.

Yes, it does. The exact part that makes the controller variable available is using the additional parameter `user_ns={'controller': controller} ` into the InteractiveShellEmbed instantiation. 

For the error in closing the console, the following code should take care of it:
```
+            interactive_shell.magic("%kill_embedded -x -y")
+            interactive_shell.pt_loop.stop()

```

However, as soon as you quit Kytos, the Linux console becomes useless (i.e., the terminal encoding becomes inconsistent somehow and you cannot type any other command). To fix this, you can run the command `reset` in the Linux console.

I believe that sending a CTRL+D could work well, but I didn't find an easy way of doing that.
